drivers/hwmon_ina219.c

```diff
@@ -0,0 +1,496 @@
+			battery_voltage_max = 4350;
+		}
+		else if (d_equal(volt_nominal, 3.85)) {
+			battery_voltage_min = 3000;
+			battery_voltage_max = 4400;
+		}
+		else {
+			fatalx(EXIT_FAILURE, "unsupported battery.voltage.nominal: %lf\n",
+					volt_nominal);
+		}
+	}
+	else {
+		double tmp;
+
+		if ((ret = parse_voltage(d_min, &tmp)) < 0) {
+			errno = -ret;
+			fatal_with_errno(EXIT_FAILURE,
+					"invalid battery.voltage.low: '%s'\n", d_min);
+		}
+		else {
+			battery_voltage_min = (unsigned int) (tmp * 1000.0);
+		}
+
+		if ((ret = parse_voltage(d_max, &tmp)) < 0) {
+			errno = -ret;
+			fatal_with_errno(EXIT_FAILURE,
+					"invalid battery.voltage.high: '%s'\n", d_max);
+		}
+		else {
+			battery_voltage_max = (unsigned int) (tmp * 1000.0);
+		}
+	}
+
+	upslogx(LOG_NOTICE, "battery.voltage.low = %u mV\n", battery_voltage_min);
+	upslogx(LOG_NOTICE, "battery.voltage.high = %u mV\n", battery_voltage_max);
+}
+
+static int parse_charge(const char *s, unsigned int *v)
+{
+	long tmp;
+
+	errno = 0;
+	tmp = strtol(s, NULL, 0);
+
+	if (errno)
+		return -errno;
+
+	if (tmp < 1 || tmp > 100)
+		return -ERANGE;
+
+	*v = (int) tmp;
+	return 0;
+}
+
+static void battery_charge_params_init(void)
+{
+	const char *v_lb = getval("lb");
+	const char *v_hb = getval("hb");
+	int ret;
+
+	if (v_lb) {
+		if ((ret = parse_charge(v_lb, &battery_charge_low)) < 0) {
+			errno = -ret;
+			fatal_with_errno(EXIT_FAILURE,
+					"Invalid value from option lb: '%s'", v_lb);
+		}
+	}
+
+	if (v_hb) {
+		if ((ret = parse_charge(v_hb, &battery_charge_high)) < 0) {
+			fatal_with_errno(EXIT_FAILURE,
+					"Invalid value from option hb: '%s'", v_hb);
+		}
+	}
+
+	if (battery_charge_low > battery_charge_high) {
+		errno = EINVAL;
+		fatal_with_errno(EXIT_FAILURE, "Invalid options: lb > hb");
+	}
+}
+
+void upsdrv_initups(void)
+{
+	const char *hwmon_dir = SYSFS_HWMON_DIR;
+	int ret;
+
+	if (getval("sysfs_dir"))
+		hwmon_dir = getval("sysfs_dir");
+
+	if ((ret = detect_hwmon_ina219(hwmon_dir)) < 0) {
+		errno = -ret;
+		fatal_with_errno(EXIT_FAILURE, "detect_hwmon_ina219(%s) has failed",
+				hwmon_dir);
+	}
+
+	battery_voltage_params_init();
+	battery_charge_params_init();
+
+	upslogx(LOG_NOTICE, "lb = %u hb = %u\n", battery_charge_low, battery_charge_high);
+}
+
+void upsdrv_initinfo(void)
+{
+	dstate_setinfo("ups.mfr", "%s", "Texas Instruments");
+	dstate_setinfo("ups.model", "%s", "INA219");
+	dstate_setinfo("ups.type", "%s", "ups");
+	dstate_setinfo("device.mfr", "%s", "Texas Instruments");
+	dstate_setinfo("device.model", "%s", "INA219");
+	dstate_setinfo("device.type", "%s", "ups");
+	dstate_setinfo("device.description", "%s",
+			"Bidirectional Current/Power Monitor With I2C Interface");
+
+	dstate_setinfo("battery.charge.low", "%d", battery_charge_low);
+	dstate_setinfo("battery.charge.high", "%d", battery_charge_high);
+}
+
+static unsigned int battery_charge_compute(void)
+{
+	const unsigned int divisor =
+		(battery_voltage_max - battery_voltage_min) / 100;
```

Comment:
I wonder if rounding to `int` so early is problematic for precision. In practice I'd expect this to vary between roughly 10 to 15 or so (e.g. `(4200-3000)/100 = 12.0`), so every whole unit may be significant in terms of needless error induction.

I'd suggest changing `divisor` and `charge` to `double` and only cast to `unsigned int` in the end.

Reply:
Done.